spring-integration-core/src/main/java/org/springframework/integration/graph/GraphBuilder.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.graph;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.SmartInitializingSingleton;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.integration.channel.AbstractMessageChannel;
+import org.springframework.integration.endpoint.EventDrivenConsumer;
+import org.springframework.integration.endpoint.IntegrationConsumer;
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.integration.endpoint.PollingConsumer;
+import org.springframework.integration.endpoint.SourcePollingChannelAdapter;
+import org.springframework.integration.gateway.MessagingGatewaySupport;
+
+/**
+ * Builds the runtime object model graph.
+ *
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public class GraphBuilder implements ApplicationContextAware, SmartInitializingSingleton {
+
+	private ApplicationContext applicationContext;
+
+	private Graph graph;
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;
+	}
+
+	public Graph getGraph() {
+		return this.graph;
+	}
+
+	@Override
+	public void afterSingletonsInstantiated() {
+		Map<String, AbstractMessageChannel> channels = this.applicationContext
+				.getBeansOfType(AbstractMessageChannel.class);
+		Map<String, SourcePollingChannelAdapter> spcas = this.applicationContext
+				.getBeansOfType(SourcePollingChannelAdapter.class);
+		Map<String, MessagingGatewaySupport> gateways = this.applicationContext
+				.getBeansOfType(MessagingGatewaySupport.class);
+		Map<String, MessageProducerSupport> producers = this.applicationContext
+				.getBeansOfType(MessageProducerSupport.class);
+		Map<String, PollingConsumer> polling = this.applicationContext.getBeansOfType(PollingConsumer.class);
+		Map<String, EventDrivenConsumer> eventDriven = this.applicationContext
+				.getBeansOfType(EventDrivenConsumer.class);
```

Comment:
Since you have made already `IntegrationConsumer extends NamedComponent` this two are redundant and can be replaced just only with one: `this.applicationContext.getBeansOfType(IntegrationConsumer.class)`